src/api/ApiProvider.js

```diff
@@ -0,0 +1,304 @@
+        const user = doc(this.users, userID);
+        return (await getDoc(user)).data();
+    }
+
+    /**
+     * @callback getUserCallback
+     * @param {import('./types').User} user
+     */
+    /**
+     * @param {string} userID
+     * @param {getUserCallback} cb
+     */
+    trackGetUser(userID, cb) {
+        const user = doc(this.users, userID);
+        return onSnapshot(user, (d) => cb(d.data()));
+    }
+
+    /**
+     * @param {string} userID
+     */
+    async getUserPosts(userID) {
+        const q = query(this.posts, where('authorID', '==', userID));
+        const posts = await Promise.all(
+            (
+                await getDocs(q)
+            ).docs.map(async (d) => this.getPost(d.id, d.data()))
+        );
+        return posts;
+    }
+
+    /**
+     * @callback getUserPostsCallback
+     * @param {import('./types').PostFull[]} posts
+     */
+    /**
+     * @param {string} userID
+     * @param {getUserPostsCallback} cb
+     */
+    trackGetUserPosts(userID, cb) {
+        const q = query(this.posts, where('authorID', '==', userID));
+        return onSnapshot(q, async (d) =>
+            cb(
+                await Promise.all(
+                    d.docs.map(async (e) => this.getPost(e.id, e.data()))
+                )
+            )
+        );
+    }
+
+    /**
+     * @param {string} userID
+     * @param {import('./types').User} userData
+     */
+    async setUser(userID, userData) {
+        const user = doc(this.users, userID);
+        return setDoc(user, userData);
+    }
+
+    /**
+     * @param {string} userID
+     */
+    async getUserChatRooms(userID) {
+        const q = query(
+            this.chat_rooms,
+            where('members', 'array-contains', userID)
+        );
+        return Promise.all(
+            (await getDocs(q)).docs.map(async (d) =>
+                this.getChatRoom(d.id, d.data())
+            )
+        );
+    }
+
+    /**
+     * @callback getUserChatRoomsCallback
+     * @param {import('./types').ChatRoomFull[]} rooms
+     */
+    /**
+     * @param {string} userID
+     * @param {getUserChatRoomsCallback} cb
+     */
+    trackGetUserChatRooms(userID, cb) {
+        const q = query(
+            this.chat_rooms,
+            where('members', 'array-contains', userID)
+        );
+        return onSnapshot(q, async (d) =>
+            cb(
+                await Promise.all(
+                    d.docs.map(async (e) => this.getChatRoom(e.id, e.data()))
+                )
+            )
+        );
+    }
+
+    /**
+     * @param {string} chatRoomID
+     * @param {import('./types').ChatRoom} alreadyFetchedData
+     * @returns {import('./types').ChatRoomFull}
+     */
+    async getChatRoom(chatRoomID, alreadyFetchedData) {
+        /**
+         * @type {import('./types').ChatRoom}
+         */
+        // eslint-disable-next-line operator-linebreak
+        const chat_room =
+            // eslint-disable-next-line operator-linebreak
```

Comment:
disabling eslint for own convenience is baad

Reply:
In this case it didn't cooperate well with Prettier. I don't know what's going on with this eslint setup but it just doesn't like prettier.

Reply:
Fixed!